core/include/vecmem/memory/impl/atomic_ref.ipp

```diff
@@ -0,0 +1,184 @@
+/*
+ * VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+#pragma once
+
+// SYCL include(s).
+#if defined(CL_SYCL_LANGUAGE_VERSION) || defined(SYCL_LANGUAGE_VERSION)
+#include <CL/sycl.hpp>
+#endif
+
+/// Helpers for explicit calls to the SYCL atomic functions
+#if defined(CL_SYCL_LANGUAGE_VERSION) || defined(SYCL_LANGUAGE_VERSION)
+#define __VECMEM_SYCL_ATOMIC_CALL0(FNAME, PTR) \
+    cl::sycl::atomic_##FNAME<value_type>(      \
+        cl::sycl::atomic<value_type>(cl::sycl::global_ptr<value_type>(PTR)))
+#define __VECMEM_SYCL_ATOMIC_CALL1(FNAME, PTR, ARG1)                         \
+    cl::sycl::atomic_##FNAME<value_type>(                                    \
+        cl::sycl::atomic<value_type>(cl::sycl::global_ptr<value_type>(PTR)), \
+        ARG1)
+#define __VECMEM_SYCL_ATOMIC_CALL2(FNAME, PTR, ARG1, ARG2)                   \
+    cl::sycl::atomic_##FNAME<value_type>(                                    \
+        cl::sycl::atomic<value_type>(cl::sycl::global_ptr<value_type>(PTR)), \
+        ARG1, ARG2)
+#endif
+
+namespace vecmem {
+
+template <typename T>
+VECMEM_HOST_AND_DEVICE atomic_ref<T>::atomic_ref(reference ref) : m_ptr(&ref) {}
+
+template <typename T>
+VECMEM_HOST_AND_DEVICE void atomic_ref<T>::store(value_type data) {
+
+#if (defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__)) && \
+    (!defined(SYCL_LANGUAGE_VERSION))
+    volatile pointer addr = m_ptr;
+    __threadfence();
+    *addr = data;
+#elif defined(CL_SYCL_LANGUAGE_VERSION) || defined(SYCL_LANGUAGE_VERSION)
+    __VECMEM_SYCL_ATOMIC_CALL1(store, m_ptr, data);
+#else
+    *m_ptr = data;
+#endif
+}
+
+template <typename T>
+VECMEM_HOST_AND_DEVICE auto atomic_ref<T>::load() const -> value_type {
+
+#if (defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__)) && \
+    (!defined(SYCL_LANGUAGE_VERSION))
+    volatile pointer addr = m_ptr;
+    __threadfence();
+    const value_type value = *addr;
+    __threadfence();
+    return value;
+#elif defined(CL_SYCL_LANGUAGE_VERSION) || defined(SYCL_LANGUAGE_VERSION)
+    return __VECMEM_SYCL_ATOMIC_CALL0(load, m_ptr);
+#else
+    return *m_ptr;
+#endif
+}
+
+template <typename T>
+VECMEM_HOST_AND_DEVICE auto atomic_ref<T>::exchange(value_type data)
+    -> value_type {
+
+#if (defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__)) && \
+    (!defined(SYCL_LANGUAGE_VERSION))
+    return atomicExch(m_ptr, data);
+#elif defined(CL_SYCL_LANGUAGE_VERSION) || defined(SYCL_LANGUAGE_VERSION)
+    return __VECMEM_SYCL_ATOMIC_CALL1(exchange, m_ptr, data);
+#else
+    value_type current_value = *m_ptr;
+    *m_ptr = data;
+    return current_value;
```

Comment:
I don't want to add those kind of compiler specific hacks. :frowning:

But I also don't want to prevent us writing code that would test the behaviour of things like `vecmem::device_vector` in host code. (Which would be impossible to do if we'd need to wait for C\+\+20.)

This is why `vecmem::atomic` does what it does since the beginning as well. We do not assume at the moment that you'd want to write (for instance) OpenMP code that would happily fill a "device vector". For that we will indeed need to wait for C\+\+20.

Reply:
Mmm... on one hand, I see where you are coming from, on the other hand, I find it unfortunate from an API ergonomics standpoint that vecmem has a type that is named like a (future) std type, but does not behave like said std type in some circumstances (host code).

Perhaps a way to sidestep the issue would be to adjust the name a bit in order to clarify that this is not quite `std::atomic_ref`, but rather a type which only behaves like `std::atomic_ref` in device code. How would you feel about renaming this type to `std::device_atomic_ref`?

Reply:
Fine, `vecmem::device_atomic_ref` is not a too bad name. (I was considering `vecmem::atomic_device_ref` for a moment, but `vecmem::device_atomic_ref` is a bit more in line with the other "device classes" that we have already.)